main.rb

```diff
@@ -0,0 +1,154 @@
+# rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+
+module Enumerable
+  # 1. my_each method
+
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    array = to_a
+    size.times { |index| yield array[index] }
+    self
+  end
+
+  # 2. my_each_with_index
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    arr = to_a
+    pos = 0
+    while pos < arr.length
+      yield(arr[pos], pos)
+      pos += 1
+    end
+
+    self
+  end
+
+  # 3.my_select
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    new_arr = []
+    to_a.my_each { |item| new_arr << item if yield(item) }
+    new_arr
+  end
+
+  # 4.my_all?
+
+  def my_all?(args = nil)
+    if block_given?
+      counter_false = 0
+      my_each { |num| counter_false += 1 unless yield num }
+      counter_false.zero?
+    elsif args.nil
+      my_all? { |num| num }
+    else
+      my_all? { |num| args == num }
+    end
+  end
```

Comment:
my_all?

when no block or argument is given returns true when none of the collection members are false or nil:
`p [1, true, 'hi', []].my_all?` should return true.

when a class is passed as an argument returns true if all of the collection is a member of such class
`p [1,1,1].my_all?(Integer)` should return true

when a Regex is passed as an argument returns true if all of the collection matches the Regex
`p ['d','d'].my_all?(/d/)` should return true

when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
`p [3,3,3,3].my_all?(3)` should return true